src/System.Linq/src/System/Linq/Single.cs

```diff
@@ -149,5 +150,79 @@ public static TSource SingleOrDefault<TSource>(this IEnumerable<TSource> source,
 
             return default(TSource);
         }
+
+        public static bool TrySingle<TSource>(this IEnumerable<TSource> source, out TSource element)
+        {
+            if (source == null)
```

Comment:
Why not just have all of the logic in a private TrySingle that returns a tri-state rather than a bool?